### PR TITLE
Add option to mute boosts from a user (#648)

### PR DESCRIPTION
**Pull request by spinda**
This implements #648, adding an additional relationship allowing users to mute particular users' boosts without muting them entirely. I've found myself wanting this quite a bit, as some of the users I follow are a bit over-eager with the boost button.

A new menu item is added to user profiles:

user profile menu screenshot

A menu item is also added to boosted statuses, to mute boosts from the user doing the boosting:

status menu screenshot

This is my first time contributing to Mastodon, or even writing Ruby, so please let me know where I've messed up. 🌈

### CODE REVIEW

**Comment by spinda** [comment]
Documentation: https://github.com/tootsuite/documentation/pull/260

File: app/javascript/mastodon/components/account.js [hunk 1]
```diff
@@ -46,6 +48,10 @@ class Account extends ImmutablePureComponent {
     this.props.onMute(this.props.account);
   }
 
+  handleMuteBoosts () {
```
Comment (ykzts) [inline app/javascript/mastodon/components/account.js]: `handleMuteBoosts = () => { /* ... */ }`

**Review by ykzts — COMMENTED** [review]
It may be better to name the internal process "MuteReblog" instead of "MuteBoost". Internally the word "Reblog" is used until now.

**Comment by spinda** [comment]
@ykzts I've applied the requested changes and rebased against the latest `master`.

File: db/schema.rb [hunk 2]
```diff
@@ -71,6 +71,14 @@
...
   end
 
+  create_table "boosts_mutes", force: :cascade do |t|
+    t.integer  "account_id",        null: false
+    t.integer  "target_account_id", null: false
+    t.datetime "created_at",        null: false
+    t.datetime "updated_at",        null: false
+    t.index ["account_id", "target_account_id"], name: "index_boosts_mutes_on_account_id_and_target_account_id", unique: true, using: :btree
+  end
+
```
Comment (ykzts) [inline db/schema.rb]: unused?

File: app/javascript/mastodon/locales/defaultMessages.json [hunk 3]
```diff
@@ -1131,4 +1163,4 @@
     ],
     "path": "app/javascript/mastodon/features/ui/components/video_modal.json"
   }
-]
\ No newline at end of file
+]
```
Comment (ykzts) [inline app/javascript/mastodon/locales/defaultMessages.json]: did you run `yarn manage:translations`?

File: db/schema.rb (hunk 2 above)
Comment (spinda) [inline db/schema.rb]: I was just mimicking the schema for `mutes` here. Will take it out.

File: app/javascript/mastodon/locales/defaultMessages.json (hunk 3 above)
Comment (spinda) [inline app/javascript/mastodon/locales/defaultMessages.json]: I did not, that's very convenient. It should probably be added to the development docs.

**Comment by spinda** [comment]
@ykzts I've made the requested changes.

File: db/schema.rb (hunk 2 above)
Comment (ykzts) [inline db/schema.rb]: I'm sorry for lack of a word. You don't need to delete timestamps.

Delete the `boosts_mutes` table.

**Review by akihikodaki — APPROVED** [review]

File: db/schema.rb [hunk 4]
```diff
@@ -71,6 +71,12 @@
     t.index ["account_id", "target_account_id"], name: "index_blocks_on_account_id_and_target_account_id", unique: true, using: :btree
   end
 
+  create_table "boosts_mutes", force: :cascade do |t|
+    t.integer  "account_id",        null: false
+    t.integer  "target_account_id", null: false
+    t.index ["account_id", "target_account_id"], name: "index_boosts_mutes_on_account_id_and_target_account_id", unique: true, using: :btree
+  end
```
Comment (ykzts) [inline db/schema.rb]: Please delete this `create_table`.

File: db/migrate/20170525214746_create_reblogs_mutes.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class CreateReblogsMutes < ActiveRecord::Migration[5.0]
+  def change
+    create_table :reblogs_mutes do |t|
+      t.integer :account_id, null: false
+      t.integer :target_account_id, null: false
```
Comment (ykzts) [inline db/migrate/20170525214746_create_reblogs_mutes.rb]: Add `t.timestamps`.

**Review by ykzts — CHANGES_REQUESTED** [review]

**Comment by Gargron** [comment]
Have you considered whether it'd be worth it to store this on the follows table, as a kind-of "option" of following someone? I am not saying that's definitely the better choice, but want to raise the possibility.

Pros: no extra table, no extra queries to get that value

Cons: follows table is usually huge, modifying it could take a long time, muted reblogs might be a rarely enough used feature that adding a boolean for *all* follows might be worse in terms of disk usage

**Comment by ykzts** [comment]
@spinda ping

**Comment by nightpool** [comment]
@Gargron if it's a nullable value, shouldn't that be very quick to add to the table?

**Comment by Gargron** [comment]
@nightpool Booleans shouldn't be nullable!

**Comment by Gargron** [comment]
@ekiru @aschmitz Anyone want to revive this PR in the same style as optional notification muting?

**Comment by aschmitz** [comment]
Well, not to stomp over this PR, but I did just write up glitch-soc/mastodon#212 for the same feature with much the same UI accidentally. I can make that a PR here too, if you like.

**Pull request closed without merging**